tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -25,41 +25,79 @@
-  } else if (symbol == "float2") {
-    const uint32_t floatType = theBuilder.getFloatType();
-    return theBuilder.getVec2Type(floatType);
-  } else if (auto *builtinType = dyn_cast<BuiltinType>(type.getTypePtr())) {
-    switch (builtinType->getKind()) {
-    case BuiltinType::Void:
-      return theBuilder.getVoidType();
-    case BuiltinType::Float:
-      return theBuilder.getFloatType();
-    default:
-      // TODO: handle other primitive types
-      assert(false && "unhandled builtin type");
-      break;
+/// SPIR-V type instructions generated during translation will be emitted to
+/// the SPIR-V module builder passed into the constructor.
+/// Warnings and errors during the translation will be reported to the
+/// DiagnosticEngine passed into the constructor.
+class TypeTranslator {
+public:
+  TypeTranslator(spirv::ModuleBuilder &builder, DiagnosticsEngine &diag)
+      : theBuilder(builder), diags(diag) {}
+
+  /// \brief Generates the corresponding SPIR-V type for the given Clang
+  /// frontend type and returns the type's <result-id>. On failure, reports
+  /// the error and returns 0.
+  ///
+  /// The translation is recursive; all the types that the target type depends
+  /// on will be generated.
+  uint32_t translateType(QualType type) {
+    const auto *typePtr = type.getTypePtr();
+
+    // Primitive types
+    if (const auto *builtinType = dyn_cast<BuiltinType>(typePtr)) {
+      switch (builtinType->getKind()) {
+      case BuiltinType::Void:
+        return theBuilder.getVoidType();
+      case BuiltinType::Float:
+        return theBuilder.getFloatType();
+      default:
+        emitError("Primitive type '%0' is not supported yet.")
+            << builtinType->getTypeClassName();
+        return 0;
+      }
+    }
+
+    // In AST, vector types are TypedefType of TemplateSpecializationType.
+    // We handle them via HLSL type inspection functions.
+    if (hlsl::IsHLSLVecType(type)) {
+      const auto elemType = hlsl::GetHLSLVecElementType(type);
+      const auto elemCount = hlsl::GetHLSLVecSize(type);
+      return theBuilder.getVecType(translateType(elemType), elemCount);
+    }
+
+    // Struct type
+    if (const auto *structType = dyn_cast<RecordType>(typePtr)) {
+      const auto *decl = structType->getDecl();
+
+      // Collect all fields' types.
+      std::vector<uint32_t> fieldTypes;
+      for (const auto *field : decl->fields()) {
+        fieldTypes.push_back(translateType(field->getType()));
+      }
+
+      return theBuilder.getStructType(fieldTypes);
     }
-  } else {
-    // TODO: handle other types
-    assert(false && "unhandled clang type");
+
+    emitError("Type '%0' is not supported yet.") << type->getTypeClassName();
+    return 0;
   }
-  return 0;
-}
+
+private:
+  /// \brief Wrapper method to create an error message and report it
+  /// in the diagnostic engine associated with this consumer.
+  template <unsigned N> DiagnosticBuilder emitError(const char (&message)[N]) {
```

Comment:
Not really since they both refer to class variables. We can do some macro stuff, but still needs wrap around. I won't bother that much for two instances. If it ends up we need more, we should solve this.